cf-custom-resources/lib/cert-replicator.js

```diff
@@ -0,0 +1,324 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+/* jshint node: true */
+/* jshint esversion: 8 */
+
+"use strict";
+
+const aws = require("aws-sdk");
+
+const defaultSleep = function (ms) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+};
+
+// These are used for test purposes only
+let defaultResponseURL;
+let defaultLogGroup;
+let defaultLogStream;
+let waiter;
+let sleep = defaultSleep;
+let maxAttempts = 10;
+
+/**
+ * Upload a CloudFormation response object to S3.
+ *
+ * @param {object} event the Lambda event payload received by the handler function
+ * @param {object} context the Lambda context received by the handler function
+ * @param {string} responseStatus the response status, either 'SUCCESS' or 'FAILED'
+ * @param {string} physicalResourceId CloudFormation physical resource ID
+ * @param {object} [responseData] arbitrary response data object
+ * @param {string} [reason] reason for failure, if any, to convey to the user
+ * @returns {Promise} Promise that is resolved on success, or rejected on connection error or HTTP error response
+ */
+let report = function (
+  event,
+  context,
+  responseStatus,
+  physicalResourceId,
+  responseData,
+  reason
+) {
+  return new Promise((resolve, reject) => {
+    const https = require("https");
+    const { URL } = require("url");
+
+    let responseBody = JSON.stringify({
+      Status: responseStatus,
+      Reason: reason,
+      PhysicalResourceId: physicalResourceId || context.logStreamName,
```

Comment:
Do you think we can:
```suggestion
      PhysicalResourceId: physicalResourceId,
```

Reply:
this is to make the error msg slightly more helpful. because at least they'll be easier to find out where to look for the lambda logs, instead of just an empty string.

Reply:
The log stream name is already in the `reason` though! I'm just not super sure if we should set `PhysicalResourceId` to a log stream name but judging from how we handle `Delete` it seems to be okay.